### PR TITLE
Modify VTGenerator to work with Hotspot

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
Modified the ValueType Generator so that all tests pass using the
hotspot JVM.
- Move init helper to ref types only
- Classfile version is 56 instead of 55
- Remove monitor enter and monitor exit tests as hotpsot enforces
structured locking. Only montiorEnterAndExit remains.
- Modified wither and getter methods
- Test syncMethod disabled temporarily due to lack of hotspot support
on ValueTypes by appropriate exception handling.
- checkCast on invalid Qtype throws NPE

Modified ValueTypeTests:
- disabled com.ibm.jvm.Dump.SystemDump() due to incompatibility with
hotspot
- changed all instances of generate(Getter | Wither) to
generateGeneric(Getter | Wither)

Disabled following tests due to incompatible code on OpenJ9:
- testDefaultValueWithNonValueType
- testCheckCastValueTypeOnInvalidQtype

Signed-off-by: Oussama Saoudi <oussama.saoudi@ibm.com>

### CODE REVIEW

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 1]
```diff
@@ -1173,32 +1148,29 @@ static public void testMonitorEnterOnValueType() throws Throwable {
...
-			monitorExitOnValueType.invoke(valueTypeArray);
+			monitorEnterAndExitOnValueType.invoke(valueTypeArray);
 		} catch (IllegalMonitorStateException e) {
 			Assert.fail("Should not throw exception. MonitorExit can be used with ValueType arrays");
 		}
 	}
 
 
-	@Test(priority=2)
+	@Test(groups="openj9-only", priority=2)
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:1173]: Actually, maybe just disable this `enabled=false`

File: test/functional/Valhalla/playlist.xml [hunk 2]
```diff
@@ -25,6 +25,7 @@
 	<test>
 		<testCaseName>ValueTypeTests</testCaseName>
 		<variations>
+			<variation>NoOptions</variation>
```
Comment (tajila) [inline test/functional/Valhalla/playlist.xml:28]: I dont think this mode will work for J9

File: test/functional/Valhalla/playlist.xml [hunk 3]
```diff
@@ -49,6 +50,7 @@
 			<subset>Valhalla</subset>
 		</subsets>
 		<impls>
+			<impl>hotspot</impl>
```
Comment (tajila) [inline test/functional/Valhalla/playlist.xml:53]: I'll try to figure out a new way to do this. For now, being able to run with hotspot manually is good enough

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by fjeremic** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
@OussamaSaoudi There are some NoSuchMethodExceptions

**Comment by OussamaSaoudi** [comment]
Ah right, we'll have to change how openj9 works since we have changed the expected method signatures. All tests pass on hotspot save for IdentityInterface ones since the interface is not defined there.
The primary change I expect is how we determine Q or L type on method signatures. You'll notice that many of the changes are simply: L -> Q or vis versa (with some of them requiring getSigFromSimpleName instead of hard-coding one or the other).

Apologies, it's been a while since I've looked at this.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeGenerator.java [hunk 4]
```diff
@@ -463,10 +446,10 @@ private static void makeValue(ClassWriter cw, String valueName, String makeValue
 		mv.visitEnd();
 	}
 	
-	private static void makeValueTypeDefaultValue(ClassWriter cw, String valueName, String makeValueSig, String[] fields, int makeMaxLocal) {
+	private static void makeValueTypeDefaultValue(ClassWriter cw, String valueName, boolean isRef, String makeValueSig, String[] fields, int makeMaxLocal) {
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeGenerator.java:449]: can please add the new field to the end, just for consistency

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 5]
```diff
@@ -3035,12 +2984,12 @@ static public void testUnresolvedPutFieldUse() throws Throwable {
 	static public void testIdentityObjectOnValueType() throws Throwable {
 		String fields[] = {"longField:J"};
 		Class valueClass = ValueTypeGenerator.generateValueClass("testIdentityObjectOnValueType", fields);
-		assertFalse(Arrays.asList(valueClass.getInterfaces()).contains(IdentityInterface.class));
+		//assertFalse(Arrays.asList(valueClass.getInterfaces()).contains(IdentityInterface.class));
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:2987]: you can leave these in for now, we will change the name in another PR

Comment (tajila) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:2987]: keep the tests enabled, we will need a second pass at this. Next items will be to rename the IndentityInterface and add MH support for VTs. After we can try this again. For now the fixes you have are fine, just need to make sure the tests keep running.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 6]
```diff
@@ -2550,7 +2499,7 @@ static public void testCheckCastValueTypeOnInvalidLtype() throws Throwable {
 	/*
 	 * Ensure that casting null to invalid Qtype class will throw a NoClassDef 
 	 */
-	@Test(priority=1, expectedExceptions=NoClassDefFoundError.class)
+	@Test(priority=1, expectedExceptions=NullPointerException.class)
```
Comment (OussamaSaoudi) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:2502]: @tajila Heads up this test is still throwing NoClassDefFoundError, so is failing.

Comment (OussamaSaoudi) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:2502]: testDefaultValueWithNonValueType is also failing due to OpenJ9 and Hotspot expecting different signature to defaultvalue(). Lmk how to proceed on both of them.

Comment (tajila) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:2502]: You can disable this one. For each disabled test please open an issue describing what the RI behaviour is and how we differ. Also add me to it.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java [hunk 7]
```diff
@@ -3035,12 +2984,12 @@ static public void testUnresolvedPutFieldUse() throws Throwable {
 	static public void testIdentityObjectOnValueType() throws Throwable {
```
Comment (tajila) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:2984]: Leave these tests as is, same as below. We will need more changes to get this working with the RI. These changes will be done in other PRs

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Pull request merged**